ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java

```diff
@@ -85,17 +92,27 @@ public NetworkClient(Selector selector, NetworkConfig networkConfig, int maxConn
    */
   public List<ResponseInfo> sendAndPoll(List<RequestInfo> requestInfos)
       throws IOException {
-    if (closed) {
-      throw new IllegalStateException("The NetworkClient is closed.");
-    }
-    List<ResponseInfo> responseInfoList = new ArrayList<ResponseInfo>();
-    for (RequestInfo requestInfo : requestInfos) {
-      pendingRequests.add(new RequestMetadata(time.milliseconds(), requestInfo));
+    long startTime = time.milliseconds();
+    try {
+      if (closed) {
+        throw new IllegalStateException("The NetworkClient is closed.");
+      }
+      List<ResponseInfo> responseInfoList = new ArrayList<ResponseInfo>();
+      numPendingConnections.set(pendingRequests.size());
+      for (RequestInfo requestInfo : requestInfos) {
+        ClientNetworkRequestMetrics clientNetworkRequestMetrics =
+            new ClientNetworkRequestMetrics(networkMetrics.requestQueueTime, networkMetrics.requestSendTime,
+                networkMetrics.requestSendTotalTime, 0);
+        pendingRequests.add(new RequestMetadata(time.milliseconds(), requestInfo, clientNetworkRequestMetrics));
+        numPendingConnections.set(pendingRequests.size());
+      }
+      List<NetworkSend> sends = prepareSends(responseInfoList);
+      selector.poll(POLL_TIMEOUT_MS, sends);
+      handleSelectorEvents(responseInfoList);
+      return responseInfoList;
+    } finally {
+      networkMetrics.networkClientSendAndPollTime.update(time.milliseconds() - startTime);
```

Comment:
I think you should set `numPendingConnections` (which is really `numPendingRequests`) just once - and here in finally. The `NetworkClient` adds all requests to the pending list and then they get a chance to be sent out only when `prepareSends` is called. So these requests are really pending, as far as the metrics are concerned, only if they remain in the pending list even after `prepareSends()` was called.